lib/archive/unpack_unix.go

```diff
@@ -131,71 +137,104 @@ func InvokeUnpack(op trace.Operation, tarStream io.Reader, filter *FilterSpec, r
 	return nil
 }
 
-// Unpack hooks into a binary present in the appliance vm called unpack in order to execute InvokeUnpack inside of a chroot. This method works identically to InvokeUnpack, except that it will not function in areas where the binary is not present at /bin/unpack
-func Unpack(op trace.Operation, tarStream io.Reader, filter *FilterSpec, root string) error {
+// OfflineUnpack wraps Unpack for usage in contexts without a childReaper.
+func OfflineUnpack(op trace.Operation, tarStream io.Reader, filter *FilterSpec, root string) error {
+
+	var cmd *exec.Cmd
+	var err error
+	if cmd, err = Unpack(op, tarStream, filter, root, ApplianceBinaryPath); err != nil {
+		return err
+	}
+
+	if err = cmd.Wait(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Unpack runs the binary compiled in cmd/unpack.go which creates a chroot at `root` and passes `op`, `tarStream`, and `filter` to InvokeUnpack for extraction of the tar on the filesystem. `binPath` should be either ApplianceBinaryPath or ContainerBinaryPath. Unpack returns a `Cmd` to allow use in conjunction with the tether's `LaunchUtility`, so it is necessary to call `cmd.Wait` after `Unpack` exits e.g. OfflineUnpack, if not being used in conjunction with LaunchUtility and the childReaper.
+func Unpack(op trace.Operation, tarStream io.Reader, filter *FilterSpec, root string, binPath string) (*exec.Cmd, error) {
 
 	fi, err := os.Stat(root)
 	if err != nil {
 		// the target unpack path does not exist. We should not get here.
 		op.Errorf("tar unpack target does not exist: %s", root)
-		return err
+		return nil, err
 	}
 
 	if !fi.IsDir() {
 		err := fmt.Errorf("unpack root target is not a directory: %s", root)
 		op.Error(err)
-		return err
+		return nil, err
 	}
 
 	encodedFilter, err := EncodeFilterSpec(op, filter)
 	if err != nil {
 		op.Error(err)
-		return err
+		return nil, err
 	}
 
 	// Prepare to launch the binary, which will create a chroot at root and then invoke InvokeUnpack
 	// #nosec: Subprocess launching with variable. -- neither variable is user input & both are bounded inputs so this is fine
-	cmd := exec.Command("/bin/unpack", op.ID(), root, *encodedFilter)
+	// "/bin/unpack" on appliance
+	// "/.tether/unpack" inside container
+	cmd := exec.Command(binPath, op.ID(), root, *encodedFilter)
 
-	//stdin
 	stdin, err := cmd.StdinPipe()
 
 	if err != nil {
 		op.Error(err)
-		return err
+		return cmd, err
 	}
 
 	if stdin == nil {
 		err = errors.New("stdin was nil")
 		op.Error(err)
-		return err
+		return cmd, err
 	}
-	done := make(chan error)
+
 	go func() {
-		// copy the tarStream to the binary via stdin; the binary will stream it to InvokeUnpack unchanged
-		defer stdin.Close()
+		if tr, ok := tarStream.(*tar.Reader); ok {
+			tw := tar.NewWriter(stdin)
+			defer stdin.Close()
```

Comment:
Do you really want to close stdin only if it is a tar.Reader?

Reply:
it gets explicitly closed when we see an EOF, further down (there's a `tw.Close()`), so sometimes this call will be redundant and throw an (uncaptured) error, but this is just to make sure that the stream closes regardless as to whether the tar routine below succeeds. That said, I should probably `defer tw.Close()` instead of `stdin.Close()`. The difference in this context, though, I believe to be fairly minor. I don't believe there's an effect on functionality, between the two.